### ISSUE TITLE
Non-ASCII char support for Windows terminals

### ISSUE THREAD

**Issue by Eskibear** [microsoft/vscode-java-debug#622]
When launching in integrated/external terminal, the debugger passes commands to vscode, and vscode assembles and sends it to the terminal. But if there's non-ascii char in code/path, the java process will have unexpected behavior, because of bad Unicode support in Windows. 

Either VS Code or the extension should do something to cover these cases. 

## Example
Folder structure:
```bash
中文目录
├── Hello.class
└── Hello.java
0 directories, 2 files
```

Content of file `中文目录/Hello.java`:
```java
public class Hello {
  public static void main(String[] args) {
    System.out.println("中文");
  }
}
```

### a) on Mac OSX
Works as expected
```bash
$> java -cp 中文目录 Hello
中文
```
#### Non-ASCII char in classpath
Classpath is not correctly recognized by Java, if it contains non-ascii char.
```cmd
D:\Test>java -cp 中文目录 Hello
Error: Could not find or load main class Hello
```

#### Non-ASCII char in stdout
cd into the folder, the classpath is correctly recognized now, but the terminal fails to print non-ascii chars. 
```cmd
D:\Test>cd 中文目录

D:\Test\中文目录>java -cp . Hello
ä??æ??
```
To display non-ascii char in Windows terminals, you have to modify the `code page`, see https://ss64.com/nt/chcp.html

Workaround: set code page to 65001 (utf-8)
```cmd
D:\Test\中文目录>chcp 65001
Active code page: 65001

D:\Test\中文目录>java -cp . Hello
中文
```

**Comment by Eskibear** [comment]
Currently in v0.21.0 we use a `launcher.bat` to set code page to 65001 before launching the java process. 
But it's reported to have issues:
* If IT Policy forbid to run .bat files, debugging is totally blocked. #646 
* It doesn't work well with _Git Bash for Windows_ for some users. #642 
* For MacOS, commands sent to integrated terminal sometimes is incomplete, thus debuggee process is not started, leading to debugger timeout. #614 #651

**Comment by Eskibear** [comment]
> Non-ASCII char in classpath
> Classpath is not correctly recognized by Java, if it contains non-ascii char.
> ```
> D:\Test>java -cp 中文目录 Hello
> Error: Could not find or load main class Hello
> ```

Tried in 1.51.1, above issue has been fixed, but non-ascii stdout is still not working well. see below.


![image](https://user-images.githubusercontent.com/2351748/99049957-6f8fff80-25d2-11eb-8a66-ed96906082f6.png)

**Comment by testforstephen** [comment]
@Eskibear Have a loot at the screenshot above, Hello.java file is using UTF-8, but the javac and java command is using the default platform encoding, which is Chinese encoding. That's why Chinese character is not printed well.

**Comment by Eskibear** [comment]
@testforstephen Yes, as shown in my screenshot, it's cp936 (Chinese GB2312 encoding).  The solutions can be either:
* change file encoding to GB2312 (cp936).
* change code page of the terminal to 65001 (UTF-8).

So I don't think it's an issue of Java debugger extension. It's due to inconsistent encoding between .java file and terminal, which also results in bad output even outside VS Code. I think we can close this one now.

**Comment by NinelieZ** [comment]
In my situation, stdout works well, but it works wrong when i trying to input a Chinese string with Java Extension Pack.
![image](https://user-images.githubusercontent.com/44182970/106694590-d6dab000-6613-11eb-98a1-932116467f8f.png)

**Comment by imba-tjd** [comment]
I'd like to discuss this issue.

First of all, according to my test, C C++ C# java can't read Chinese characters (actually including all characters beyond ASCII) under chcp 65001.

The example you showed above works for me under 936.

![图片](https://user-images.githubusercontent.com/24759802/120071319-50197980-c0c1-11eb-836b-dfe574124111.png)

Do you have a good reason to use chcp 65001?

**Comment by Eskibear** [comment]
> Do you have a good reason to use chcp 65001?

Use UTF8 for better compatibility. Inconsistent encoding can happen among System(Windows), terminal(PS), JVM, editor (VSCode)...etc.

If you have issue with that, you can open `C:\Users\<username>\.vscode\extensions\vscjava.vscode-java-debug-<version>\scripts\launcher.bat` and change code page back to 936. Let me know if that works for all your use cases, e.g.
- print Chinese chars
- input Chinese chars in terminal (CMD/PS)
- work with path containing Chinese chars
-

**Comment by imba-tjd** [comment]
Let me show you how the program works when using CLI. The following tests are not related to `launcher.bat` and `vscode-java-debug`. The Code Page is 936.

One way is using GBK as the file encoding.

![图片](https://user-images.githubusercontent.com/24759802/120149588-8586aa00-c21c-11eb-831a-e894d789788c.png)

When using UTF8 file encoding, simply use `javac -encoding utf8`.

![图片](https://user-images.githubusercontent.com/24759802/120149878-e9a96e00-c21c-11eb-9041-a0e79868096e.png)

But there is no way to work using `java test.java` directly.

![图片](https://user-images.githubusercontent.com/24759802/120150167-4ad14180-c21d-11eb-825f-4e0fc26b8027.png)

The reason is that `file.encoding` by default is corresponding to CP 936 GBK. When not set `-Dfile.encoding` or set it to GBK, it tries to read source code using GBK, thus causes **hard code** string messy (`package 目录;`). But the input/output are normal:

![图片](https://user-images.githubusercontent.com/24759802/120150782-0f834280-c21e-11eb-89ef-794ba50973a5.png)

When using `-Dfile.encoding=UTF8`, hard code string is normal, but runtime encoding is also set to utf8 and the CP is 936 so it can't input or output.

Idealy chcp 65001 would solve this, because both source and runtime are utf8. But due to bugs in Windows console, programs can't read things in 65001.

So my conclusion is that this extension should compile source code with `-encoding utf8` and do not set `-Dfile.encoding` to UTF8.

**Comment by Eskibear** [comment]
I met issue with your GBK example:
![image](https://user-images.githubusercontent.com/2351748/120165987-b7087100-c22e-11eb-9b55-c07224f2290c.png)

For utf8 case:
![image](https://user-images.githubusercontent.com/2351748/120166182-e9b26980-c22e-11eb-9e96-cc6ae98f650b.png)

> But there is no way to work using java test.java directly.

I don't see why you can run `java <path-to-java>` directly before compile it to a class file.
![image](https://user-images.githubusercontent.com/2351748/120166595-50d01e00-c22f-11eb-8ccc-28254fe5b982.png)

> So my conclusion is that this extension should compile source code with -encoding utf8 and do not set -Dfile.encoding to UTF8 and do not use chcp 65001.

One thing to mention, Java files are compiled by JDT language server, which I know little about the encoding it's using.  /cc @testforstephen, please share more about JDT, and see if above proposal makes any sense?

**Comment by imba-tjd** [comment]
Please retry your first two test in a new CMD&nbsp; console and in chcp 936. It's yet another issue related to powershell.


As for java test.java, it's because I'm using JDK16

**Comment by testforstephen** [comment]
> @Eskibear Java files are compiled by JDT language server, which I know little about the encoding it's using.

Java language server uses the encoding from VS Code setting `"files.encoding"` to compile your code.

> @imba-tjd in your first screenshot
![image](https://user-images.githubusercontent.com/44182970/106695042-bbbc7000-6614-11eb-9f57-0ff78a7e3ba1.png)

The encoding is set to utf-8 on our Java tool level, but it still doesn't display well. We will investigate to find the gaps.

**Comment by imba-tjd** [comment]
`javac` doesn't recognize files.encoding, it only uses `-encoding`.

Then, as I said before, you must compile in utf8 (if the source code is utf8) and **run in gbk** and chcp must always be 936.

**Comment by testforstephen** [comment]
> Then, as I said before, you must compile in utf8 (if the source code is utf8) and run in gbk and chcp must always be 936.

This is not a systematic way to solve the problem because this approach doesn't display the string literals in your code well.

encoding in launch.json only influences the run-time. Compile is controlled by Java language server, which reads the encoding from VS Code setting "files.encoding" to compile your code.

**Comment by imba-tjd** [comment]
> this approach doesn't display the string literals in your code well

…………… I didn't post that blue sceenshot

> reads the encoding from VS Code setting "files.encoding" to compile your code.

Thanks, I think so! So the compile phase is working.

---

Then with these two changes, there is no messy code:

* remove chcp 65001 in `launcher.bat`
* add `"encoding": "GBK"` in `launch.json`

![图片](https://user-images.githubusercontent.com/24759802/120267730-22a61900-c2d7-11eb-8543-1641150ff767.png)

However, I can't hit debug point when path contains Chinese (but the program itself runs).

So the issue becomes: make this extension work, under `-Dfile.encoding=GBK` and `chcp 936`.
Because this envrionment is what `java xxx` runs on. When you not set `-Dfile.encoding`, the default value is corresponding to chcp, i.e. 936

**Comment by Eskibear** [comment]
### TL;DR
- I agree that it's better not to force 65001, if we find ways to cover all these non-ascii cases.
- To correctly print String literals, `file.encoding` should match code page.
- To correctly input strings, developers should ensure terminal is using the same encoding/codepage with charset used by Scanner.


### Compiling 
As long as javac/JDT uses the same encoding as the file is written in, it should work. Otherwise it would probably fail to compile. Let's skip discussing about it.

### String Literal
First of all, it should be compiled with the same encoding as written in. (otherwise it's impossible to print it correctly).
**If `file.encoding` matches code page, it's printing correct string literals.** 


My expriment:
Literal.java is encoded in utf8, simply printing a string.
![image](https://user-images.githubusercontent.com/2351748/120273182-bc25f880-c2e0-11eb-89cc-df2fbc681e79.png)
same in CMD
![image](https://user-images.githubusercontent.com/2351748/120273766-83d2ea00-c2e1-11eb-80d5-10701b3fd842.png)
![image](https://user-images.githubusercontent.com/2351748/120273753-80d7f980-c2e1-11eb-9ab2-8d5f4becdd6d.png)

It's strange to me that, when code page is 936, it works for all above `file.encoding`. I don't know if it's trick from JVM/Windows.

### String from I/O

I don't think it's a big problem, and it should be managed by users own. For `Scanner`, you can initialize it with a charset instead of leaving it empty. If charset is not specified, it fallbacks to use default charset `file.encoding` (in my JVM its default value is cp1252).

**As long as charset used by Scanner matches terminal's code page, it should work.** Thus it doesn't have to be related with `file.encoding`

My experiments below:
- `IO.java` prints raw bytes in hex. It first prints "你好" in both GBK and UTF-8, and then prints bytes from your input. **Note that I forced the scanner to use GBK**.
```
Scanner scanner = new Scanner(System.in, "GBK");
```


![image](https://user-images.githubusercontent.com/2351748/120276180-f2657700-c2e4-11eb-9b9d-da35c87a62e9.png)


- `Echo.java` is a program simply prints what you input.

![image](https://user-images.githubusercontent.com/2351748/120277541-c21ed800-c2e6-11eb-8b2a-2e6d05a16827.png)

**Comment by Eskibear** [comment]
As far as I understand now, maybe we should apply below changes:
- remove `launcher.bat` (no more `chcp 65001`), and verify if there's potential issues.
- provide better guide for non-ascii users, e.g. set encoding in launch.json, etc.

**Comment by imba-tjd** [comment]
String Literal: I agree.

provide better guide: Give them a notice not using non-English path.

IO:

> As long as charset used by Scanner matches terminal's code page, it should work.

Not true. You tested GBK in Scanner. This test uses all UTF8:

![图片](https://user-images.githubusercontent.com/24759802/120280017-bda7ee80-c2e9-11eb-9a8d-d907082da774.png)
